File: back/src/domain/use_cases/insert_salary.rs
use crate::domain::models::{Captcha, Salary};
use crate::infra::captcha_service::ValidateError;
use crate::infra::CaptchaService;
use crate::infra::{salary_repository::InsertError, SalaryRepository};
use std::sync::Arc;

pub enum Error {
    Unknown(&'static str),
}

pub async fn insert_salary(
    service: Arc<dyn CaptchaService>,
    repo: Arc<dyn SalaryRepository>,
    captcha: Captcha,
    salary: Salary,
) -> Result<(), Error> {
    match service.validate(captcha).await {
        Ok(_) => {}
        Err(ValidateError::Unknown(str)) => return Err(Error::Unknown(str)),
    };

    match repo.insert(salary).await {
        Ok(_) => Ok(()),
        Err(InsertError::Unknown(str)) => Err(Error::Unknown(str)),
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::infra::{InMemorySalaryRepository, StubCaptchaService};

    #[tokio::test]
    async fn it_should_return_an_unknown_error_when_the_captcha_is_invalid() {
        let captcha = Captcha::test();
        let salary = Salary::test();
        let service = Arc::new(StubCaptchaService::new().with_error());
        let repo = Arc::new(InMemorySalaryRepository::new());

        let res = insert_salary(service, repo, captcha, salary).await;

        match res {
            Err(Error::Unknown(_)) => {}
            _ => unreachable!(),
        };
    }

    #[tokio::test]
    async fn it_should_return_an_unknown_error_when_the_salary_could_not_be_inserted() {
        let captcha = Captcha::test();
        let salary = Salary::test();
        let service = Arc::new(StubCaptchaService::new());
        let repo = Arc::new(InMemorySalaryRepository::new().with_error());

        let res = insert_salary(service, repo, captcha, salary).await;

        match res {
            Err(Error::Unknown(_)) => {}
            _ => unreachable!(),
        };
    }

    #[tokio::test]
    async fn it_should_return_ok_otherwise() {
        let captcha = Captcha::test();
        let salary = Salary::test();
        let service = Arc::new(StubCaptchaService::new());
        let repo = Arc::new(InMemorySalaryRepository::new());

        let res = insert_salary(service, repo, captcha, salary).await;

        match res {
            Ok(()) => {}
            _ => unreachable!(),
        };
    }
}

File: src/domain/lib/insertSalary.ts
import {
    CaptchaService, 
    SalaryRepository, 
    TokenRepository,
    TokenSender
} from "@infra";
import { Result } from "@utils";
import { Captcha, Salary, Token } from "../models/mod.ts";

export async function insertSalary(
    captchaService: CaptchaService,
    salaryRepository: SalaryRepository,
    tokenRepository: TokenRepository,
    tokenSender: TokenSender,
    captcha: Captcha,
    salary: Salary,
): Promise<Result<void, string>> {
    const validationResult = await captchaService.validate(captcha);
    if (Result.isErr(validationResult)) { return validationResult; }

    const insertionResult = await salaryRepository.insert(salary);
    if (Result.isErr(insertionResult)) { return insertionResult; }

    const token = Token.generate();

    const sendingResult = await tokenSender.send(token, salary.email);
    if (Result.isErr(sendingResult)) { return sendingResult; }

    return tokenRepository.insert(salary.id, token);
}
